modules/jod-ohjaaja-cms-navigation/src/main/java/fi/okm/jod/ohjaaja/cms/navigation/service/NavigationServiceImpl.java

```diff
@@ -0,0 +1,305 @@
+import com.liferay.site.navigation.service.SiteNavigationMenuItemService;
+import com.liferay.site.navigation.service.SiteNavigationMenuService;
+import com.liferay.site.navigation.util.comparator.SiteNavigationMenuItemOrderComparator;
+import fi.okm.jod.ohjaaja.cms.navigation.dto.NavigationDto;
+import fi.okm.jod.ohjaaja.cms.navigation.dto.NavigationItemDto;
+import fi.okm.jod.ohjaaja.cms.navigation.rest.application.NavigationRestApplication;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+@Component(immediate = true, service = NavigationService.class)
+public class NavigationServiceImpl implements NavigationService {
+
+  private static final String ASSET_CATEGORY_CLASS_NAME =
+      "com.liferay.asset.kernel.model.AssetCategory";
+  private static final String JOURNAL_ARTICLE_CLASS_NAME =
+      "com.liferay.journal.model.JournalArticle";
+
+  private static final String SAP_ENTRY_NAME = "JOD_OHJAAJA_NAVIGATION";
+  private static final Map<Locale, String> SAP_DESCRIPTION =
+      Map.of(LocaleUtil.fromLanguageId("en_US"), "Public access JOD OHJAAJA NAVIGATION");
+  private static final String SERVICE_SIGNATURE = NavigationRestApplication.class.getName() + "#*";
+
+  private static final String CUSTOM_FIELD_NAME = "Type";
+  private static final String[] CUSTOM_FIELD_DEFAULT_DATA =
+      new String[] {"Article", "CategoryListing", "CategoryMain"};
+
+  private static final Log log = LogFactoryUtil.getLog(NavigationServiceImpl.class);
+
+  @Reference private SiteNavigationMenuItemService siteNavigationMenuItemService;
+
+  @Reference private SiteNavigationMenuService siteNavigationMenuService;
+
+  @Reference private AssetCategoryService assetCategoryService;
+
+  @Reference private JournalArticleService journalArticleService;
+
+  @Reference private JournalArticleResourceLocalService journalArticleResourceLocalService;
+
+  @Reference private SAPEntryLocalService sapEntryLocalService;
+
+  public NavigationDto getNavigation(Long siteId, String languageId) {
+    List<SiteNavigationMenu> siteNavigationMenus =
+        siteNavigationMenuService.getSiteNavigationMenus(20117);
+
+    if (!siteNavigationMenus.isEmpty()) {
+      return toNavigationDto(siteNavigationMenus.getFirst(), languageId);
+    }
+
+    return null;
+  }
+
+  @Override
+  public void initNavigation() {
+    try {
+      sapEntryLocalService.getSAPEntry(PortalUtil.getDefaultCompanyId(), SAP_ENTRY_NAME);
+    } catch (PortalException portalException) {
+      initServiceAccessPolicy();
+    }
+    initCustomField();
+  }
+
+  private NavigationDto toNavigationDto(SiteNavigationMenu siteNavigationMenu, String languageId) {
+    Map<Long, List<SiteNavigationMenuItem>> siteNavigationMenuItemsMap =
```

Comment:
matter of style, but preferring `var` where possible would make the code bit more succinct (e.g. here the type is already clear from the return type of the helper function).